rfcs/0044-ckb-light-client/0044-ckb-light-client.md

```diff
@@ -0,0 +1,519 @@
+node, with following data:
+- The blocks hash of the tip block that client knows.
+- The blocks hash of the start block.
+- The blocks number of the start block.
+- The number $L$.
+- The difficulty boundary $D_{\mathrm{boundary}}$.
+- The difficulty samples. If there are no difficulties sampled in the
+  previous step, leave them to be empty.
+
+At last, the server replies the proof of the sampled blocks.
+
+Accordingly, the client can treat that tip block as a valid tip block after
+checking the returned proof.
+
+#### Phase #2: Client asks a proof for some blocks.
+
+- At the start, the client sends the `GetBlocksProof` message, which
+  contains the last block hash it knows, and hashes of the blocks which
+  require verifying, to the server which has the best chain.
+
+- Then, the server should reply a `SendBlocksProof` message.
+
+  If the last block that the client known is in the server's current chain,
+  then the server should return
+  - the verifiable header of the last block that the client known,
+  - a proof for blocks which are from the provided block hashes and in the
+    current chain,
+  - the headers of those proved blocks,
+  - hashes of blocks which are from the provided block hashes but not in the
+    current chain.
+
+  Otherwise, the server should only return its current verifiable tip
+  header.
+
+#### Phase #3: Client asks a proof for some transactions.
+
+- At the start, the client sends the `GetTransactionsProof` message, which
+  contains the last block hash it knows, and hashes of the transactions
+  which require verifying, to the server which has the best chain.
+
+- Then, the server should reply a `SendTransactionsProof` message.
+
+  If the last block that the client known is in the server's current chain,
+  then the server should return
+  - the verifiable header of the last block that the client known,
+  - a proof for blocks which contain valid transactions from the provided
+    transaction hashes,
+  - the proofs that proving those valid transactions are in those blocks,
+  - hashes of transactions which are from the provided transaction hashes
+    but not in the current chain.
+
+  Otherwise, the server should only return its current verifiable tip
+  header.
+
+## Limitations
+
+To avoid attacks by malicious clients, there are few limitations in server
+side.
+- In `GetLastStateProof` messages, the sum of the size of difficulties and 2
+  times of the `last_n_blocks` should NOT greater than 1000.
+- In `GetBlocksProof` messages, the size of block hashes should NOT greater
+  than 1000.
+- In `GetTransactionsProof` messages, the size of transactions hashes should
+  NOT greater than 1000.
+
+There are also few limitations in messages to improvement performance.
+- In `GetLastStateProof` messages, the difficulties should be sorted.
+- In `SendLastStateProof` messages, the headers should be sorted.
+
+##### How a Server Choose Blocks from Sampled Difficulties?
+
+## Deployment
+
```

Comment:
Adds testnet activation parameters. The parameters are:
start =  5_346 epoch,    2022-10-31 23:00:00 UTC
timeout=  5_616 epoch,     2022-12-14 23:00:00 UTC
min_activation_epoch= 5_676,  2022-12-24 23:00:00 UTC
threshold: 75%, (testnet only)

Reply:
mention This RFC is deployed concurrently with 0045.

Reply:
For CKB mainnet and testnet, this RFC will deploy by "CKB softfork activation" with the name "LightClient" and bit 1, using [0043](https://github.com/nervosnetwork/rfcs/blob/master/rfcs/0043-ckb-softfork-activation/0043-ckb-softfork-activation.md) modified to use the cellbase witness field instead of the block header version for signaling.

Reply:
For CKB testnet, the start is epoch 5346 (Approximately 2022-10-31 15:00:00 UTC), timeout is epoch timestamp 5616 (Approximately 2022-10-31 15:00:00 UTC), the threshold 75%, and the min_activation_epoch is epoch 5676 (Approximately 2022-12-24 15:00:00 UTC).